test/test_ops.py

```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+                             exact_device=False,
+                             **kwargs)
+        else:
+            self.assertEqual(actual, expected, msg, exact_device=False, **kwargs)
+
+    def _test_reference_numerics(self, dtype, op, tensors, equal_nan=True):
+        def _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan=True):
+            if not torch.can_cast(numpy_to_torch_dtype_dict[expected.dtype.type], dtype):
+                exact_dtype = False
+
+            if dtype in [torch.uint8, torch.int8, torch.bool]:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=1e-3, atol=1e-2)
+
+            elif dtype is torch.bfloat16:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=16e-3, atol=1e-5)
+
+            else:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+        for sample_input in tensors:
+            sample_args = sample_input.args
+            sample_kwargs = sample_input.kwargs
+            sample_input = sample_input.input
+            torch_kwargs, numpy_kwargs = op.sample_kwargs(sample_input.device, dtype, sample_input)
+            numpy_args = []
+            if dtype is torch.bfloat16:
+                for sample_arg in sample_args:
+                    numpy_arg = sample_arg.cpu().to(torch.float32).numpy() if isinstance(sample_arg, torch.Tensor) \
+                                    else sample_arg
+                    numpy_args.append(numpy_arg)
+                sample_numpy_input = sample_input.cpu().clone().to(torch.float32).numpy()
+            else:
+                for sample_arg in sample_args:
+                    numpy_arg = sample_arg.cpu().numpy() if isinstance(sample_arg, torch.Tensor) else sample_arg
+                    numpy_args.append(numpy_arg)
+                sample_numpy_input = sample_input.cpu().clone().numpy()
+            actual = op(sample_input, *sample_args, **torch_kwargs)
+            expected = op.ref(sample_numpy_input, *numpy_args, **numpy_kwargs)
+
+            if sample_input.numel() < 10:
+                msg = ("Failed to produce expected results! Input tensor was"
+                        " {0}, torch result is {1}, and reference result is"
+                        " {2}.").format(sample_input, actual, expected)
+            else:
+                msg = None
+
+            exact_dtype = True
+            if isinstance(actual, torch.Tensor):
+                _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan)
+            else:
+                for x, y in zip(expected, actual):
+                    _helper_reference_numerics(x, y, msg, exact_dtype, equal_nan)
+
+    @suppress_warnings
+    @ops(only_testing_ops)
+    def test_reference_testing(self, device, dtype, op):
```

Comment:
If this PR updates the SampleInput class and adds a function common_utils for comparing with a NumPy reference, this function might look like:

```
@suppress_warnings
@ops(only_testing_ops, allowed_dtypes=(torch.float32, torch.long))
def test_reference_testing(self, device, dtype, op):
  for sample in op.sample_inputs(device, dtype):
    compare_with_numpy(op, op.ref, sample)
```

And then in common_utils.py the compare with NumPy function might look like:

```
def compare_with_numpy(op, ref, sample):
  sample_np = sample.numpy()
  actual = op(sample.splat())
  expected = op(sample.splat())
  self.assertEqual(actual, expected)
```

I cheated by also adding a `splat` method to sample inputs to help avoid having to parse out the input, arg, and kwargs all the time.

What would you think of a structure like this, @krshrimali, @kshitij12345?

Reply:
- I like the idea of having `compare_with_numpy` function and keeping the test neat and simple.
- Minor correction, I think you meant `expected = op.ref(sample_np.splat())` (just mentioning if someone reads this in future).

For `splat()` function, do you think it's better to have a static method `splat()` in the `SampleInput` class? The reason I say this is because `<SampleInputObject>.numpy()` should return a tuple of `(inputs, args, kwargs)` and not a `SampleInput` object. Hence it will not be possible to access `sample_np.splat()` (`smaple_np` being a tuple).

```python
actual = op(SampleInput.splat(sample))
expected = op.ref(SampleInput.splat(sample_np))
```

Having a splat function which takes an argument instead of `self`, is probably more useful? (please correct me if I'm wrong here)

```python ​
@staticmethd
def splat(sample):
   # inputs, args, kwargs (order)
   ​return tuple(sample[0], sample[1], sample[2],)
```

Instead of:

```python
# This will work if .numpy() returns a SampleInput object
def splat(self):
    return tuple(self.args, self.input, self.kwargs,)
```

_Note:_ the functions are just here for example, the actual implementation will be more accurate and different.

What are your thoughts on this? @mruberry @pmeier @kshitij12345

Reply:
That's a great point. What about a `splat()` method that splats and a `numpy()` method that converts tensors into NumPy arrays but return a new SampleInput? Couldn't that be splatted?

Reply:
I did think of that as a solution but I am not sure if a splat function will work like what we expect it to be. Here is a workaround though:

```python
import torch

class SampleInput:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c

    def fn(self, a):
        print(a)

    def splat(self, fn):
        return fn(self.a, self.b, **self.c,)

obj = SampleInput(torch.tensor(1), torch.tensor(2), {'alpha': 2})
print(obj.splat(torch.add))  # Returns tensor(5)
```

What do you think about this structure? @mruberry - Personally, I think it's reasonable as it does make things easier.

Reply:
Seems like the idea of splat() isn't very interesting and we should just drop it from this PR